### ISSUE TITLE
Repeated bad readings with hlw8012

### ISSUE THREAD

**Issue by Lovely-Maisonette** [esphome/issues#2699]
### The problem

There is one wrong power value, coming several times per minute when noting is connected to the powerplug.
Flipping (A+V) or not flipping (Just V) the sel pin makes no difference.
I need to filter one power value that comes wrongly several times per minute (always the same 2.92998 value, not just noise)
I plugged the powerplug into a quality powermeter, which shows 0.00W and 0.00A

I also get 0.15A - 0.21A values instead of zero on the A/V pin (in case this info is relevant) so I decided to ignore Amps (not sure how to fix this, using a linear filter can give negative values but never 0) for the moment, since the chip provides amps info with a different pin, and is not used to calculate power (power is what I need).

I get a mostly linear output once I start cranking up the wattage (halogen light with restive dimmer + hairdrier.
It deviates proportionally, ESPHOME is ca. 10% higher, except only at around 570W, the only area where ESPHome shows LOWER values. I assume that is something inherent to the sensor, but better to share it, in case it uncovers another issue.

### Which version of ESPHome has the issue?

v 2021.10.3

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

latest / docker, but irrelevant

### What platform are you using?

ESP8266

### Board

https://templates.blakadder.com/bakibo_TP22Y.html

### Component causing the issue

hlw8012

### Example YAML snippet

```yaml
sensor:
  - platform: hlw8012
    model: BL0937
    cf_pin: ${PM_CF}
    cf1_pin: ${PM_CF1}
    sel_pin:  
      number: ${PM_Sel}
      inverted: ${PM_Seli}
    voltage_divider: 1700
    change_mode_every: 4294967295   # I get a 0.2A reading when nothing connected, ignoring for the moment
    initial_mode: VOLTAGE
    update_interval: 1s

    power:
      id: Power1
      name: "${nice_name} Power"
      unit_of_measurement: W
      accuracy_decimals: 1
      filters:
        - filter_out: 2.92998
```

### Anything in the logs that might be useful for us?

```txt
[00:49:46][D][sensor:113]: 'powerplug2 Power': Sending state 0.00000 W with 1 decimals of accuracy
[00:49:47][D][hlw8012:082]: Got power=0.0W, voltage=232.1V
[00:49:47][D][sensor:113]: 'powerplug2 Power': Sending state 0.00000 W with 1 decimals of accuracy
[00:49:48][D][hlw8012:082]: Got power=0.0W, voltage=230.5V
[00:49:48][D][sensor:113]: 'powerplug2 Power': Sending state 0.00000 W with 1 decimals of accuracy
[00:49:49][D][hlw8012:082]: Got power=2.9W, voltage=230.5V
[00:49:50][D][hlw8012:082]: Got power=0.0W, voltage=230.2V
[00:49:50][D][sensor:113]: 'powerplug2 Power': Sending state 0.00000 W with 1 decimals of accuracy
[00:49:51][D][hlw8012:082]: Got power=0.0W, voltage=230.9V
[00:49:51][D][sensor:113]: 'powerplug2 Power': Sending state 0.00000 W with 1 decimals of accuracy
```

**Comment by Lovely-Maisonette** [comment]
With this filter I get perfect values all the way up to 250W (the strongest halogen I can dim at the moment)
```
      filters:
        - filter_out: 2.92998
        - delta: 0.1
        - multiply: 0.935
```
Below 50W is marginally less linear, I will use the linear filter to improve that at some point in time.
Is the filter order relevant? I would need to first filter out 2.92998 and only after apply a linear filter that starts with 0 -> 0

**Comment by Lovely-Maisonette** [comment]
Additionally, Energy keeps growing even it Power is 0 (after filter)
```
[02:55:11][D][hlw8012:082]: Got power=0.0W, voltage=232.1V
[02:55:11][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01465 Wh with 1 decimals of accuracy
[02:55:12][D][hlw8012:082]: Got power=0.0W, voltage=232.4V
[02:55:12][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01465 Wh with 1 decimals of accuracy
[02:55:13][D][hlw8012:082]: Got power=0.0W, voltage=231.8V
[02:55:13][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01465 Wh with 1 decimals of accuracy
[02:55:14][D][hlw8012:082]: Got power=0.0W, voltage=232.2V
[02:55:14][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01506 Wh with 1 decimals of accuracy
[02:55:15][D][hlw8012:082]: Got power=0.0W, voltage=232.4V
[02:55:15][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01506 Wh with 1 decimals of accuracy
[02:55:16][D][hlw8012:082]: Got power=0.0W, voltage=232.0V
[02:55:16][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01506 Wh with 1 decimals of accuracy
[02:55:17][D][hlw8012:082]: Got power=0.0W, voltage=232.1V
[02:55:17][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01506 Wh with 1 decimals of accuracy
[02:55:18][D][hlw8012:082]: Got power=0.0W, voltage=232.2V
[02:55:18][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01506 Wh with 1 decimals of accuracy
[02:55:19][D][hlw8012:082]: Got power=0.0W, voltage=232.1V
[02:55:19][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01506 Wh with 1 decimals of accuracy
[02:55:20][D][hlw8012:082]: Got power=0.0W, voltage=232.7V
[02:55:20][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01506 Wh with 1 decimals of accuracy
[02:55:21][D][hlw8012:082]: Got power=0.0W, voltage=231.8V
[02:55:21][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01546 Wh with 1 decimals of accuracy
[02:55:22][D][hlw8012:082]: Got power=0.0W, voltage=232.2V
[02:55:22][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01546 Wh with 1 decimals of accuracy
[02:55:23][D][hlw8012:082]: Got power=0.0W, voltage=232.2V
[02:55:23][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01546 Wh with 1 decimals of accuracy
[02:55:24][D][hlw8012:082]: Got power=0.0W, voltage=232.1V
[02:55:24][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01546 Wh with 1 decimals of accuracy
[02:55:25][D][hlw8012:082]: Got power=0.0W, voltage=232.0V
[02:55:25][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01587 Wh with 1 decimals of accuracy
[02:55:26][D][hlw8012:082]: Got power=0.0W, voltage=232.2V
[02:55:26][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01587 Wh with 1 decimals of accuracy
[02:55:27][D][hlw8012:082]: Got power=0.0W, voltage=232.0V
[02:55:27][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01628 Wh with 1 decimals of accuracy
[02:55:28][D][hlw8012:082]: Got power=0.0W, voltage=232.4V
[02:55:28][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01668 Wh with 1 decimals of accuracy
[02:55:29][D][hlw8012:082]: Got power=0.0W, voltage=232.1V
[02:55:29][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01668 Wh with 1 decimals of accuracy
[02:55:30][D][hlw8012:082]: Got power=0.0W, voltage=232.1V
[02:55:30][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01668 Wh with 1 decimals of accuracy
[02:55:31][D][hlw8012:082]: Got power=0.0W, voltage=232.0V
[02:55:31][D][sensor:113]: 'powerplug2 Energy': Sending state 0.01709 Wh with 1 decimals of accuracy
[02:55:32][D][hlw8012:082]: Got power=0.0W, voltage=232.5V
```
Is Energy an internal ESPHome calculation? The chip does not provide that into. For which period? I guess it is picking up the unfiltered Wattage.
Is it a duplicate of platform: total_daily_energy, but inside the sensor?

**Comment by Lovely-Maisonette** [comment]
Why did this get close?
